### PR TITLE
find mkl and tbb from oneApi automatically.

### PR DESCRIPTION
**Pull request by wisherhxl**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
Thank you for the contribution!

There is already similar patch.
Could you please validate changes from here: #19384 ?
Let us know if it works for you. Thank you!

**Comment by wisherhxl** [comment]
> Thank you for the contribution!
> 
> There is already similar patch.
> Could you please validate changes from here: #19384 ?
> Let us know if it works for you. Thank you!

Thanks for the quick reply.
I've noticed the patch #19384, it's main purpose is to make OpenCV compile and run correctly with oneTBB, and I've tested on my computer with oneApi installed, it can't find both tbb and mkl automatically.
So I write this patch, which can find tbb and mkl from oneApi automatically.

I think both can be useful, As I'm new here, can you please give me a hint on what to do in this situation?

**Comment by alalek** [comment]
Please provide your use case.

Validated use cases are described here: https://github.com/opencv/opencv/pull/19384#issuecomment-769611514

**Comment by wisherhxl** [comment]
Validated with:
* [x]  Intel oneAPI Base Toolkit 2021.1 (on Windows 10):
oneAPI on windows will add ENV{ONEAPI_ROOT} automatically, so the following command can be run directly (cmd or gui).
```
cmake ../opencv -DWITH_TBB=ON -DCMAKE_DISABLE_FIND_PACKAGE_TBB=1 -DMKL_WITH_TBB=ON
...
-- Found TBB (oneApi): D:/Program Files (x86)/Intel/oneAPI/tbb/latest
...
-- Found MKL 2021.0.1 at: D:/Program Files (x86)/Intel/oneAPI/mkl/latest
-- LAPACK(MKL): LAPACK_LIBRARIES: D:/Program Files (x86)/Intel/oneAPI/mkl/latest/lib/intel64/mkl_intel_lp64.lib;D:/Program Files (x86)/Intel/oneAPI/mkl/latest/lib/intel64/mkl_sequential.lib;D:/Program Files (x86)/Intel/oneAPI/mkl/latest/lib/intel64/mkl_core.lib
-- LAPACK(MKL): Support is enabled.
```

**Comment by wisherhxl** [comment]
I've compiled the merged version again on windows, succeeded!
And the error log of the unsuccessful build "OpenCV CN Windows 10 x64" shows no code issue, so I commit it again, hopefully it can pass the build test.

Error log from the unsuccessful build "OpenCV CN Windows 10 x64":
```
+refs/pull/*:refs/remotes/origin/pr/* # timeout=10
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from https://github.com/opencv/opencv.git
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:1001)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1242)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1302)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:125)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:93)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:80)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by asmorkalov** [comment]
@wisherhxl Thanks for the note about Windows 10 build on jenkins-ci. The Windows host has poor connectivity with external world and Git operations fail time to time due to timeout. I'm working on it right now and hopefully it'll be fixed in several days.

File: cmake/OpenCVDetectTBB.cmake [hunk 1]
```diff
@@ -57,6 +63,68 @@ function(ocv_tbb_env_verify)
...
+    endif()
+
+    set(tbb_lib_find_paths ${TBB_LIB_FIND_PATHS} ${TBB_DIR}/lib)
+    foreach(TBB_ARCH ${TBB_ARCH_LIST})
+      list(APPEND tbb_lib_find_paths
+        ${TBB_DIR}/lib/${TBB_ARCH}/vc_mt
+      )
+    endforeach()
+
+    set(mkl_lib_list "tbb" "tbb12")
```
Comment (alalek) [inline cmake/OpenCVDetectTBB.cmake:87]: 1. TBB detection file should not know anything about MKL
2. "tbb12". This doesn't look reliable: 13,14,etc
3. Adding 2 .lib files is not correct. With "tbb13" what is supposed to do?

**Review by alalek — COMMENTED** [review]
Fixed TBB detection from oneAPI on Windows is here: #19447

Fix for oneMKL with TBB may be added later.

File: cmake/OpenCVDetectTBB.cmake (hunk 1 above)
Comment (wisherhxl) [reply cmake/OpenCVDetectTBB.cmake:87]: Thanks for reviewing!

1.  Sorry, I copied the name from OpencvFindMKL.cmake and forgot  to change the name, TBB detection has no connection with MKL, I can fix this.
2 & 3. I was thinking about add all libs found, but changed my desicion after reference to mkl, I can change it to find all libs in the target directory, and also keep the distinguish between debug and release version.

I've checked #19447 which disabled auto linkage for tbb, and #19451 seems have "tbb12" problem too.

**Comment by alalek** [comment]
Please check the latest code from master branch.

**Comment by wisherhxl** [comment]
> Please check the latest code from master branch.

Every thing works fine except cmake GUI.

**Comment by alalek** [comment]
Ensure that `cmake-gui` is started from the terminal with properly activated oneAPI environment.

**Comment by wisherhxl** [comment]
> Ensure that `cmake-gui` is started from the terminal with properly activated oneAPI environment.

Put ENV{ONEAPI_ROOT} into find_path search directory can make it more convenient, cmake-gui can be started anywhere.
I can handle it, Is it necessary?

**Pull request closed without merging**